File: apps/go/internal/chapter09/bikeweather.go
package chapter09

import (
	"fmt"
	"io"
	"os"
	"slices"
	"sort"
	"strings"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter02"
	"golang.org/x/text/encoding"
	"golang.org/x/text/encoding/japanese"
	"golang.org/x/text/encoding/unicode"
)

// weatherKey は自転車の表と天気の表を結合するキー。
const weatherKey = "weather_id"

// UTF8 は変換しない文字コード。bike.tsv に使う。
var UTF8 encoding.Encoding = unicode.UTF8

// ShiftJIS は weather.csv の文字コード。
var ShiftJIS encoding.Encoding = japanese.ShiftJIS

// LoadDelimited は文字コードと区切り文字を指定して、区切り文字で区切ったファイルを表に読み込む。
// 1 行目を列名として扱う。
func LoadDelimited(file string, charset encoding.Encoding, delimiter string) (chapter02.Table, error) {
	opened, err := os.Open(file)
	if err != nil {
		return chapter02.Table{}, fmt.Errorf("ファイルを開けません: %w", err)
	}

	defer func() { _ = opened.Close() }()

	content, err := io.ReadAll(charset.NewDecoder().Reader(opened))
	if err != nil {
		return chapter02.Table{}, fmt.Errorf("%s を読めません: %w", file, err)
	}

	lines := strings.Split(strings.ReplaceAll(string(content), "\r\n", "\n"), "\n")
	columns := strings.Split(strings.TrimPrefix(lines[0], "\uFEFF"), delimiter)
	rows := make([]chapter02.Row, 0, len(lines)-1)

	for _, line := range lines[1:] {
		if strings.TrimSpace(line) == "" {
			continue
		}

		values := strings.Split(line, delimiter)
		cells := make(map[string]string, len(columns))

		for i, name := range columns {
			if i < len(values) {
				cells[name] = values[i]
			} else {
				cells[name] = ""
			}
		}

		rows = append(rows, chapter02.NewRow(cells))
	}

	return chapter02.Table{Columns: columns, Rows: rows}, nil
}

// JoinWeather は天気 ID をキーにした対応表を引いて、天気の列を加える（内部結合）。
// 天気の表に無い ID の行は残さない。
func JoinWeather(bike, weather chapter02.Table) (chapter02.Table, error) {
	byID := make(map[string]chapter02.Row, len(weather.Rows))

	for _, row := range weather.Rows {
		id, err := row.Text(weatherKey)
		if err != nil {
			return chapter02.Table{}, err
		}

		byID[id] = row
	}

	added := make([]string, 0, len(weather.Columns))

	for _, column := range weather.Columns {
		if column != weatherKey {
			added = append(added, column)
		}
	}

	rows := make([]chapter02.Row, 0, len(bike.Rows))

	for _, row := range bike.Rows {
		id, err := row.Text(weatherKey)
		if err != nil {
			return chapter02.Table{}, err
		}

		found, ok := byID[id]
		if !ok {
			continue
		}

		joined, err := joinRow(row, bike.Columns, found, added)
		if err != nil {
			return chapter02.Table{}, err
		}

		rows = append(rows, joined)
	}

	columns := slices.Concat(bike.Columns, added)

	return chapter02.Table{Columns: columns, Rows: rows}, nil
}

func joinRow(row chapter02.Row, columns []string, found chapter02.Row, added []string) (chapter02.Row, error) {
	cells := make(map[string]string, len(columns)+len(added))

	for _, column := range columns {
		value, err := row.Text(column)
		if err != nil {
			return chapter02.Row{}, err
		}

		cells[column] = value
	}

	for _, column := range added {
		value, err := found.Text(column)
		if err != nil {
			return chapter02.Row{}, err
		}

		cells[column] = value
	}

	return chapter02.NewRow(cells), nil
}

// WeatherMean は天気と、その天気の日の平均利用者数。
type WeatherMean struct {
	Weather string
	Mean    float64
}

// MeanCountByWeather は天気ごとの平均利用者数を、多い順に並べて返す。
func MeanCountByWeather(joined chapter02.Table) ([]WeatherMean, error) {
	order := make([]string, 0)
	sums := make(map[string]float64)
	counts := make(map[string]int)

	for _, row := range joined.Rows {
		weather, err := row.Text("weather")
		if err != nil {
			return nil, err
		}

		count, ok, err := row.Number("cnt")
		if err != nil {
			return nil, err
		}

		if !ok {
			return nil, fmt.Errorf("利用者数が空欄です")
		}

		if _, found := sums[weather]; !found {
			order = append(order, weather)
		}

		sums[weather] += count
		counts[weather]++
	}

	means := make([]WeatherMean, 0, len(order))
	for _, weather := range order {
		means = append(means, WeatherMean{Weather: weather, Mean: sums[weather] / float64(counts[weather])})
	}

	sort.SliceStable(means, func(i, j int) bool { return means[i].Mean > means[j].Mean })

	return means, nil
}

File: apps/ruby/lib/getting_started_ml/chapter09/bike_weather.rb
# frozen_string_literal: true

require "csv"

module GettingStartedMl
  # 第 9 章の、区切り文字・文字コードを指定した読み込みと表の結合。
  module Chapter09
    # Shift_JIS のファイルを読み、UTF-8 の文字列にする指定。
    SHIFT_JIS = "Shift_JIS:UTF-8"
    # UTF-8 のファイルを読む指定。
    UTF8 = "UTF-8"
    # 結合のキーの列。
    WEATHER_KEY = "weather_id"

    module_function

    # 1 行目を列名として読み込む。文字コードが合わなければ CSV::InvalidEncodingError になる。
    def load_table(file, col_sep: ",", encoding: UTF8)
      csv = CSV.read(file, headers: true, col_sep:, encoding:)
      rows = csv.map { |record| Chapter02::Row.new(record.to_h.transform_values(&:to_s)) }

      Chapter02::Table.new(columns: csv.headers, rows:)
    end

    # 自転車の表（タブ区切り・UTF-8）を読み込む。
    def load_bike(tsv_file)
      load_table(tsv_file, col_sep: "\t")
    end

    # 天気の表（カンマ区切り・Shift_JIS）を読み込む。
    def load_weather(csv_file)
      load_table(csv_file, encoding: SHIFT_JIS)
    end

    # 天気 ID をキーにして、自転車の表に天気の列を加える（内部結合）。天気の表に無い ID の行は残さない。
    def join_weather(bike, weather)
      by_id = weather.rows.to_h { |row| [row.text(WEATHER_KEY), row] }
      rows = bike.rows.filter_map { |row| join_row(row, by_id[row.text(WEATHER_KEY)]) }

      Chapter02::Table.new(columns: bike.columns + (weather.columns - [WEATHER_KEY]), rows:)
    end

    # 自転車の行に天気の行のセルを加える。天気の行が無ければ nil を返す。
    def join_row(row, found)
      Chapter02::Row.new(row.cells.merge(found.cells.except(WEATHER_KEY))) if found
    end

    # 天気ごとの平均利用者数を、多い順に並べる。
    def mean_count_by_weather(joined)
      joined.rows.group_by { |row| row.text("weather") }
            .map { |weather, rows| [weather, rows.sum { |row| row.number("cnt") } / rows.size] }
            .sort_by { |_weather, mean| -mean }
    end
  end
end
